pkg/hive/hive.go

```diff
@@ -0,0 +1,154 @@
+	"context"
+	"fmt"
+	"time"
+
+	ma "github.com/multiformats/go-multiaddr"
+
+	"github.com/ethersphere/bee/pkg/hive/pb"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/p2p/protobuf"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+const (
+	protocolName    = "hive"
+	protocolVersion = "1.0.0"
+	peersStreamName = "peers"
+	maxPO           = 7
+	messageTimeout  = 5 * time.Second // maximum allowed time for a message to be read or written.
+)
+
+type Service struct {
+	streamer          p2p.Streamer
+	connectionManager p2p.Connecter
+	peerSuggester     DiscoveryPeerer
+	addressFinder     AddressFinder
+	logger            logging.Logger
+}
+
+type Options struct {
+	Streamer          p2p.Streamer
+	ConnectionManager p2p.Connecter
+	PeerSuggester     DiscoveryPeerer
+	AddressFinder     AddressFinder
+	Logger            logging.Logger
+}
+
+func New(o Options) *Service {
+	return &Service{
+		streamer:          o.Streamer,
+		logger:            o.Logger,
+		connectionManager: o.ConnectionManager,
+		peerSuggester:     o.PeerSuggester,
+		addressFinder:     o.AddressFinder,
+	}
+}
+
+type DiscoveryPeerer interface {
+	DiscoveryPeers(peer p2p.Peer, bin, limit int) (peers []p2p.Peer)
+}
+
+type AddressFinder interface {
+	FindAddress(overlay swarm.Address) (underlay ma.Multiaddr, err error)
+}
+
+func (s *Service) Protocol() p2p.ProtocolSpec {
+	return p2p.ProtocolSpec{
+		Name:    protocolName,
+		Version: protocolVersion,
+		Init:    s.Init,
+		StreamSpecs: []p2p.StreamSpec{
+			{
+				Name:    peersStreamName,
+				Handler: s.peersHandler,
+			},
+		},
+	}
+}
+
+// Init is called when the new peer is being initialized.
+// This should happen after overlay handshake is finished.
+func (s *Service) Init(ctx context.Context, peer p2p.Peer) error {
+	for i := 0; i < maxPO; i++ {
+		// todo: figure out the limit
+		resp, err := s.requestPeers(ctx, peer, i, 10)
+		if err != nil {
+			return err
+		}
+
+		for _, newPeer := range resp {
+			addr, err := ma.NewMultiaddr(newPeer)
+			if err != nil {
+				s.logger.Infof("Connect failed for %s: %w", newPeer, err)
+				continue
+			}
+
+			if _, err := s.connectionManager.Connect(ctx, addr); err != nil {
+				s.logger.Infof("Connect failed for %s: %w", addr.String(), err)
+				continue
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s *Service) requestPeers(ctx context.Context, peer p2p.Peer, bin, limit int) ([]string, error) {
+	stream, err := s.streamer.NewStream(ctx, peer.Address, protocolName, protocolVersion, peersStreamName)
+	if err != nil {
+		return nil, fmt.Errorf("new stream: %w", err)
+	}
+	defer stream.Close()
+
+	w, r := protobuf.NewWriterAndReader(stream)
+	if err := w.WriteMsg(&pb.GetPeers{
+		Bin:   uint32(bin),
+		Limit: uint32(limit),
+	}); err != nil {
+		return nil, fmt.Errorf("write requestPeers message: %w", err)
+	}
+
+	var peersResponse pb.Peers
+	if err := r.ReadMsgWithTimeout(messageTimeout, &peersResponse); err != nil {
```

Comment:
I think that we decided not to have a stream of peers to wait on, but to have a single request/response with multiple peers in the response. For that the timeout is just fine, as we are expecting immediate response. We do not have a loop on which we are waiting for messages.

Reply:
Exactly, I don't think it's smart to wait infinity for the response. If there is not peer to send for this been, the node should send response with no peers back, like I did in the tests for most bins.

Reply:
If we decide for the stream of peers, which we might in the future, then the msg exchange will be a bit different.

Reply:
I did not say that I'd like to see a _stream_ of peers. I was suggesting a symmetric request/reply flow. 1 request equals 1 response. Since we concluded that a stream cannot be reused this is also out of the question to have a stream of peers with the state which is represented on the protocol level.

I can definitely see the benefit in having an immediate response. It's probably a better option and will mitigate certain race conditions, so I'm fine with leaving it as it is.

Also, @pradovic please don't resolve such conversations. As mentioned in the main `swarm` repo PR [guidelines](https://github.com/ethersphere/swarm/blob/master/pull_request_template.md). Please see (6)

Reply:
@acud Oups, sorry man 🙈 I closed bunch of discussions, so I must have resolved this one by accident!
I agree, plus we avoid situation where user just does not send a message (maybe bad client or something), so we don't wait forever,

Reply:
no worries :)